Move minimum set of dynamic provisioning e2e test to testsuites

mkimuram opened the pull request:
**What this PR does / why we need it**:
This PR moves minimum set of dynamic provisioning e2e test to testsuites.
It will make all in-tree drivers do the dynamic provisioning test that CSI drivers are currently doing.

This is needed for https://github.com/kubernetes/kubernetes/issues/68024 and separated from https://github.com/kubernetes/kubernetes/pull/68025. By this separation, https://github.com/kubernetes/kubernetes/pull/68025 can focus only on making CSI drivers call the testsuites that is defined for in-tree drivers.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
/sig storage

**Release note**:
```release-note
NONE
```

msau42 commented:
Isn't dynamic provisioning already tested through the "DynamicPV" test pattern for other test suites?

mkimuram commented:
@msau42 

Thank you for your comment.

> Isn't dynamic provisioning already tested through the "DynamicPV" test pattern for other test suites?

If we only focus on the minimum test moved this time, almost equivalent test should already be done by volumeIO test with the "DynamicPV" test pattern, as you pointed out. However, we would be able to utilize other tests like pvCapacity check and ```pvCheck``` which only exists in volume_provisioning.go, in the future. (I still couldn't find a good way to define ```pvCheck``` function for each driver while keeping drivers and testsuites separate, though. Also, we might want to add tests with different provisioner parameters for each driver to this test.)

So, I think that we have two choices:
1. Not to move this test to testsuite until the test provide something more than other tests, 
2. Move this test with minimum feature this time and then improve it after that.  

Which one do you think is better?

msau42 commented:
Can we enhance `setupResource` to also validate pv capacity?

Some other functionality tested in volume_provisioning that I think may be a good candidate for the test suite:
* StorageClass mount options
* Persistence check like the "WriteReadCheck"
* Retain reclaim policy
* Block volume

mkimuram commented:
@msau42 

Thank you for your suggestion. I've updated the commit. PTAL

> Can we enhance setupResource to also validate pv capacity?

pv capacity has already been checked, because both ```ClaimSize``` and ```ExpectedSize``` are passed to testCase.

>Some other functionality tested in volume_provisioning that I think may be a good candidate for the test suite:
>    StorageClass mount options

Added. However, it seems that nfs test for this fails. (For nfs, ```StorageClass.MountOptions``` doesn't seem to affect claim spec. Instead, it seems using ```StorageClass.Parameter.MountOptions```.)
Also, note that I changed the mount option to "user" and "noauto", which should be common to all the drivers.

>    Persistence check like the "WriteReadCheck"

It has also already been checked, because the ```SkipWriteReadCheck``` is default to false.

>    Retain reclaim policy

Added.

>    Block volume

Added. However, nfs doesn't support block volume, so I just tested that it is skipped properly in my environment. It would need to be tested manually, in environment with block feature enabled.

msau42 commented:
/ok-to-test

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -0,0 +1,370 @@
+			needsCleanup = false
+			// Skip unsupported tests to avoid unnecessary resource initialization
+			skipUnsupportedTest(p, driver, pattern)
+			needsCleanup = true
```
what are you referring to by "this"? If you mean the `needsCleanup` variable I don't think we ever got to the bottom of it, the problem we ended up solving was with per-test resource cleanup racing as well as a sneaky variable shadowing issue. Those both seemed to mitigate the issue.

In terms of variable scoping for `needsCleanup` I have neither figured it out nor found anything online that specifies the behavior

davidz627 replied:
yep, that was the original hunch but it hasn't proven to be true so far. Especially after we found and fixed the other issues

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -0,0 +1,370 @@
+			if p.sc == nil {
+				framework.Skipf("Driver %q does not define Dynamic Provision StorageClass - skipping", dInfo.Name)
+			}
+			p.claimSize = "2Gi"
```
this claim size should be coming from the storage class test

davidz627 replied:
xref: https://github.com/kubernetes/kubernetes/pull/69441#discussion_r224190545

davidz627 replied:
still want this to be the claim size specified in the test struct if possible

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -0,0 +1,370 @@
+	switch volType {
+	case testpatterns.DynamicPV:
+		if dDriver, ok := driver.(drivers.DynamicPVTestDriver); ok {
+			p.sc = dDriver.GetDynamicProvisionStorageClass("")
```
do the CSI Drivers have this implemented

mkimuram replied:
Not yet. And, it is exactly what we aim to do it in https://github.com/kubernetes/kubernetes/pull/68025 .

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -0,0 +1,370 @@
+			p.claimSize = "2Gi"
+			p.pvc = getClaim(p.claimSize, ns.Name)
+			p.pvc.Spec.StorageClassName = &p.sc.Name
+			framework.Logf("In setup, pvc: %v", p.pvc)
```
This log doesn't seem accurate, `getClaim` just returns a PVC object, doesn't actually setup or do anything in the API server

mkimuram replied:
Thank you for pointing it out. It is confusing, so I will change log message.

davidz627 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -311,14 +312,14 @@ func initCSIHostpath(f *framework.Framework, config framework.VolumeTestConfig)
+		Name:         "csi-hostpath",
+		Provisioner:  "csi-hostpath",
+		Parameters:   map[string]string{},
+		ClaimSize:    "1Gi",
```
this is where claim size should be coming from right? xref: that comment i made in `setupResource`

davidz627 replied:
https://github.com/kubernetes/kubernetes/pull/69441#discussion_r224188265

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -0,0 +1,370 @@
+			resource.setupResource(driver, pattern)
+
+			// Create test input
+			input = createProvisioningTestInput(pattern, resource)
```
whats the difference between setting up a resource and creating the test input. Are they actually the same thing? I don't like how complicated this all is. We're wrapping structs in other structs then converting them to other structs through a combination of parameters and method recievers. If there's any way we could simplify all of this I would be all for it

msau42 replied:
Agree, I'm not sure if we really need an intermediate `provisioningTestResource`.  I think we can combine them all into one setup method.

mkimuram replied:
Thank you for your comment.

> whats the difference between setting up a resource and creating the test input.

"setting up a resource" is to create resources that will be used in testsuites and "creating the test input" is to put all parameters necessary to the testsuits into a single struct.

 My intention is to make ```setupResource``` sharable between testsuites, as we defined ```genericSetupResource```. Most of the testsuites will just create either (a) volumeSource or (b) object definition for creating volumeSource. So, I think that how they create the test resource is common, but which parameters in resource are required as input are different between the testsuites. (In addition, test input will require parameters other than test resource, such as driver parameters and pattern parameters.) As a result, if we combine ```setupResource``` and ```createTestInput```, we would need to implement the same logic for creating resource to all testsuites.

davidz627 replied:
can createTestInput for this test suite just call `genericSetupResource` though instead of having a seperate `setupResource` to do it? We don't have to do a huge logic change IIUC

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -0,0 +1,370 @@
+			}
+		})
+
+		testProvisioning(&input)
```
why is this a pointer?

mkimuram replied:
This is related to  https://github.com/kubernetes/kubernetes/pull/66577#discussion_r210006095 , although implementation has been changed so much since then.

In summary, we are using pattern 2 of "Global Shared Behaviors" described in https://onsi.github.io/ginkgo/#global-shared-behaviors, and it requires all the inputs for It() to be in one struct and the struct is required to be passed as a pointer.

And you are right that it should be commented in somewhere, because it is a bit tricky part and is not obvious.

davidz627 replied:
cool, thanks for the explaination! can we just add a comment above this line then

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -0,0 +1,370 @@
+	p.driver = driver
+
+	dInfo := p.driver.GetDriverInfo()
+	f := dInfo.Framework
```
Lets try to cut down on intermediate variables. When reading the code it obscures where the variable is actually coming from/being used. For example here `f` is literally only used in the very next line to get `Namespace` but because it's written here in two lines each line doesn't tell the full story and in the end its a lot of cognitive load to find out where the variable `ns` came from.

Compare this with
`ns := p.driver.GetDriverInfo().Framework.Namespace`

We can see from this line very easily that this namespace is the one defined on the framework for this driver, and that all tests sharing this driver will share this namespace (which btw doesn't seem particularly right to me as the test namespace should be per-test). 

Of course there is a time and place for intermediate variables, for example if we were to use `f` a lot of times throughout this function you don want to use `p.driver.GetDriverInfo().Framework` all the time. However, I would err on the side of clarity for readers.

Ditto this comment pretty much everywhere.

/cc @msau42 in case she disagrees

msau42 replied:
If you use a variable repeatedly, then it makes sense to create a variable for it, otherwise, I think it is fine to directly deference when you need it.

mkimuram replied:
Thank you for suggestion and detailed explanations. I will follow this guidance.

davidz627 reviewed (changes requested):
Thanks for the changes Masaki! A couple comments and questions from me.

mkimuram commented:
@davidz627 @msau42 

I fixed as suggested. PTAL

@wongma7 

In this PR, I'm adding dynamic provisioning tests with mount option specified to all drivers.
It seems that nfs external provisioner fails in this test and I found a related issue https://github.com/kubernetes-incubator/external-storage/issues/629 . 

Is my understanding right that the latest nfs-client provisioner is fixed but nfs provisioner isn't?
Then, are there any good ways to work around this issue?

[nfs-client provisioner code]
https://github.com/kubernetes-incubator/external-storage/blob/master/nfs-client/cmd/nfs-client-provisioner/provisioner.go#L83

[nfs provisioner code]
https://github.com/kubernetes-incubator/external-storage/blob/master/nfs/pkg/volume/provision.go#L224

wongma7 commented:
@mkimuram I've opened https://github.com/kubernetes-incubator/external-storage/pull/1027 to fix it. I'll push a new tag shortly, let you know when it's done.

wongma7 commented:
@mkimuram you can change this line https://github.com/kubernetes/kubernetes/blob/4278b75668c713b9a07d42de2f091b012f0e4493/test/e2e/storage/utils/utils.go#L341 to tag v2.2.0-k8s1.12 now

mkimuram commented:
@wongma7 

Thank you for quick fix! I will try it.

mkimuram commented:
@wongma7 

v2.2.0-k8s1.12 looks working fine. Thank you.

@davidz627 @msau42

PTAL 

Note that I added a code to pass driver dependent volume mount options, because supported mount options are different between drivers.

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -0,0 +1,367 @@
+
+var _ TestResource = &provisioningTestResource{}
+
+func (p *provisioningTestResource) setupResource(driver drivers.TestDriver, pattern testpatterns.TestPattern) {
```
just to confirm, this no longer needs to set up generic resource? why is that?

mkimuram replied:
@davidz627

```genericVolumeTestResource```'s ```setupResource``` will create storage class and pvc, which are the scope of provisioning test. To avoid ```setupResource``` doing too much and make test do it, ```provisioningTestResource```'s ```setupResource``` is defined here and it just creates storage class object.

davidz627 reviewed (commented):
mostly LGTM. Do you need me to test with the GCE PD CSI Driver?

mkimuram commented:
@davidz627

> mostly LGTM. 

Thank you for your review.

> Do you need me to test with the GCE PD CSI Driver?

For this PR, the change only affects in-tree drivers, so there will be no need for test with CSI drivers.
(And I confirmed that all tests for in-tree drivers including gce and aws in CI test passed.)
Once this is merged, I will work on to make this provisioning test run with csi in #68025, then your help in testing GCE PD CSI driver for the PR is really appreciated.

davidz627 commented:
/lgtm

mkimuram committed:
Move minimum set of dynamic provisioning e2e test to testsuites

mkimuram commented:
/retest

mkimuram commented:
/retest

mkimuram commented:
/retest

msau42 commented:
/lgtm
Thanks for working on this!

davidz627 commented:
@msau42 I think they decoupled lgtm and approve so you will need to do both

msau42 commented:
/approve

mkimuram commented:
/kind feature

runcom commented:
This merge is also causing test flakes in the CRI-O CI, for instance, https://openshift-gce-devel.appspot.com/build/origin-federated-results/pr-logs/pull/kubernetes-sigs_cri-o/1870/test_pull_request_crio_e2e_fedora/1368/
